tasks/windows.ps1

```diff
@@ -0,0 +1,195 @@
+  try
+  {
+    $customCACallback = {
+      param(
+        $sender,
+        [System.Security.Cryptography.X509Certificates.X509Certificate]
+        $certificate,
+        [System.Security.Cryptography.X509Certificates.X509Chain]
+        $chain,
+        [System.Net.Security.SslPolicyErrors]
+        $sslPolicyErrors
+      )
+
+      # when there are no policy errors the root certificate used to sign the
+      # masters certificate is either:
+      # * in the local trusted cert store already (other CA infra / intermediate
+      #   CA already in place)
+      # * is not self-signed / generated from a valid third-party root CA
+      if ($sslPolicyErrors -eq 'None') { return $true }
+
+      # Build a new chain that includes the Root Certificate Authority
+      $chain.ChainPolicy.ExtraStore.Add($RootCertificate)
+      $chain.Build($certificate)
+
+      # For a Puppet server response to typically be valid:
+
+      # 1. there should only be a single status
+      $oneStatus = $chain.ChainStatus.Count -eq 1
+      # 2. the chain terminated with a status of UntrustedRoot (self-signed) cert
+      # not one of the 20+ other possible failure modes
+      # https://msdn.microsoft.com/en-us/library/windows/apps/xaml/system.security.cryptography.x509certificates.x509chainstatusflags(v=vs.90).aspx
+      $untrustedRoot = [System.Security.Cryptography.X509Certificates.X509ChainStatusFlags]::UntrustedRoot
+      $untrustedCA = $chain.ChainStatus[0].Status -eq $untrustedRoot
+      # 3. the end of the chain (root) is the same CA as passed in
+      $lastCertInChain = $chain.ChainElements[$chain.ChainElements.Count - 1].Certificate
+      $rootIsExpectedCA = $lastCertInChain -eq $RootCertificate
+
+      return $oneStatus -and $untrustedCA -and $rootIsExpectedCA
+    }
+
+    # Allow install.ps1 from an endpoint signed by the Root Certificate Authority
+    [Net.ServicePointManager]::ServerCertificateValidationCallback = $customCACallback
+    $installerUri = "https://${Master}:8140/packages/current/install.ps1"
+    Write-Verbose "Downloading simplified installer $installerUri - allowing root cert [$($RootCertificate.Thumbprint)]"
+    $installer = (New-Object System.Net.WebClient).DownloadString($installerUri)
+
+    return [ScriptBlock]::Create($installer)
+  }
+  finally
+  {
+    # restore original chain validation
+    [Net.ServicePointManager]::ServerCertificateValidationCallback = $verificationCallback
+  }
+}
+
+function Out-CA($Content)
+{
+  # https://github.com/puppetlabs/puppet-specifications/blob/master/file_paths.md#puppet-agent-windows
+  $sslDirectory = Join-Path $ENV:ProgramData 'PuppetLabs\puppet\etc\ssl\certs'
+  $caFilePath = Join-Path $sslDirectory 'ca.pem'
+  New-Item -ItemType Directory -Path $sslDirectory | Out-Null
+  Write-Verbose "Writing $($Content.Length) length string to $caFilePath"
+  $Content | Out-File -FilePath $caFilePath -Encoding ASCII
+}
+
+function Invoke-SimplifiedInstaller
+(
+  $Master,
```

Comment:
Style nit: The 'method' type declaration for these helper functions could cause subtle bugs in the future, especially with the splatting use below. If any of these invocations gets parsed as an array, all the arguments will be passed as an array to the first parameter and not to all of them.

It also makes me want to cry and destroy things...so there's that.

Reply:
Yeah, I'll do that for this function, but was avoiding the extra ceremony of advanced functions for internal helpers, given there's only a single caller for any of them and there isn't a practical concern with invalid calls.